src/singleton_snap.py

```diff
@@ -0,0 +1,293 @@
+        try:
+            dir_fd = os.open(self.LOCK_DIR, os.O_RDONLY)
+            fcntl.flock(dir_fd, fcntl.LOCK_SH)
+            yield
+        except TypeError as e:
+            raise SingletonSnapError(f"Invalid lock directory path: {e}") from e
+        except FileNotFoundError as e:
+            raise SingletonSnapError(f"Lock directory not found: {e}") from e
+        except OSError as e:
+            raise SingletonSnapError(f"Error locking directory: {e}") from e
+        finally:
+            if dir_fd is not None:
+                try:
+                    fcntl.flock(dir_fd, fcntl.LOCK_UN)
+                    os.close(dir_fd)
+                except Exception as e:
+                    raise SingletonSnapError(f"Failed to release lock: {e}") from e
+
+    def _update_registration(self, snap_name: str, create: bool) -> None:
+        """Internal method to handle registration or unregistration of a snap.
+
+        Achieved by creating or removing a lock file; atomic operation is ensured
+        with a directory lock.
+
+        Raises:
+            SingletonSnapError: If an OS error occurs.
+        """
+        lock_path = self._get_registration_file_path(snap_name)
+        action = "registering" if create else "unregistering"
+
+        try:
+            with self._lock_directory():
+                if create:
+                    open(lock_path, "w").close()
+                else:
+                    try:
+                        os.remove(lock_path)
+                    except FileNotFoundError as e:
+                        # pass
+                        raise SingletonSnapError(f"Error {action} unit: {e}") from e
+        except Exception as e:
+            raise SingletonSnapError(f"Error {action} unit: {e}") from e
+
+    def get_units(self, snap_name: str) -> List[str]:
+        """Get all units currently registered for a snap (atomic with directory lock).
+
+        Args:
+            snap_name: Name of the snap to get units for
+
+        Returns:
+            List of unit names associated with the snap
+
+        Raises:
+            SingletonSnapError: If there's an error accessing the lock directory
+        """
+        prefix = f"{self.LOCK_FILE_PREFIX}{self._normalize_name(snap_name)}{self.SEPARATOR}"
+        units = []
+
+        try:
+            with self._lock_directory():
+                for filename in os.listdir(self.LOCK_DIR):
+                    if filename.startswith(prefix):
+                        units.append(filename[len(prefix) :])
+        except Exception as e:
+            raise SingletonSnapError(f"Error reading unit list: {e}") from e
+
+        return units
+
+    def is_used_by_other_units(self, snap_name: str) -> bool:
+        """Check if snap is being used by other units."""
+        return any(unit != self.unit_name for unit in self.get_units(snap_name))
+
+    @contextmanager
+    def _acquire_lock(self, lock_name: str, timeout: int) -> Generator[None, None, None]:
```

Comment:
Normally timeouts are done as float seconds in Python (which can be an int as well), so you can do 0.5 or similar.